lxd/api_cluster.go

```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
+				if err != nil {
+					return err
+				}
+
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+
+			dest, err := cluster.Connect(targetNode.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return errors.Wrap(err, "Failed to connect to destination")
+			}
+
+			source = source.UseTarget(targetNodeName)
+			source = source.UseProject(inst.Project())
+
+			req := api.InstancePost{
+				Name:      inst.Name(),
+				Migration: true,
+			}
+
+			migrationOp, err := source.MigrateInstance(inst.Name(), req)
+			if err != nil {
+				return errors.Wrap(err, i18n.G("Migration API failure"))
+			}
+
+			err = migrationOp.Wait()
+			if err != nil {
+				return err
+			}
+
+			metadata["evacuation_progress"] = fmt.Sprintf("Starting %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			startOp, err := dest.UpdateInstanceState(inst.Name(), api.InstanceStatePut{Action: "start"}, "")
+			if err != nil {
+				return err
+			}
+
+			err = startOp.Wait()
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	op, err := operations.OperationCreate(d.State(), "", operations.OperationClassTask, db.OperationClusterMemberEvacuate, nil, nil, run, nil, nil, r)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	reverter.Success()
+	return operations.OperationResponse(op)
+}
+
+func restoreClusterMember(d *Daemon, r *http.Request) response.Response {
+	originName := mux.Vars(r)["name"]
+
+	var node db.NodeInfo
+	var err error
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// Get the node.
+		node, err = tx.GetNodeByName(originName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to get node by name")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if node.State == db.ClusterMemberStatePending {
+		return response.SmartError(fmt.Errorf("Cannot restore or restore a pending node"))
+	}
+
+	if node.State == db.ClusterMemberStateCreated {
+		return response.SmartError(fmt.Errorf("Node is not evacuated"))
+	}
+
+	// The instances are retrieved in a separate transaction, after the node is in EVACUATED state.
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// If evacuating, consider only the instances on the node which needs to be evacuated.
+		dbInstances, err = tx.GetInstances(db.InstanceFilter{})
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances := make([]instance.Instance, 0)
+	localInstances := make([]instance.Instance, 0)
+
+	for _, dbInst := range dbInstances {
+		if dbInst.Node == node.Name {
+			inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+			if err != nil {
+				return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+			}
+
+			localInstances = append(localInstances, inst)
+			continue
+		}
```

Comment:
This is assuming that all instances should be running, this may not be accurate.

When shutting them down, we should override their last_state.power property to `RUNNING` and then only start back those that were running.